hydra-node/src/Hydra/Ledger/Cardano.hs

```diff
@@ -95,712 +82,510 @@ cardanoLedger =
+-- TxId
 --
 
--- TODO(AB): serialise more fields from the transaction body
-instance FromJSON CardanoTxBody where
-  parseJSON = withObject "CardanoTxBody" $ \o -> do
-    inputs <- o .: "inputs" >>= traverse parseJSON
-    outputs <- o .: "outputs" >>= traverse parseJSON
-    certificates <- o .:? "certificates" .!= mempty
-    withdrawals <- o .:? "withdrawals" .!= noWithdrawals
-    fees <- o .:? "fees" .!= mempty
-    validity <- o .:? "validity" .!= Cardano.ValidityInterval SNothing SNothing
-    auxiliaryDataHash <- o .:? "auxiliaryDataHash" .!= SNothing
-    mint <- o .:? "mint" .!= mempty
-    pure $
-      Cardano.TxBody
-        (Set.fromList inputs)
-        (StrictSeq.fromList outputs)
-        certificates
-        withdrawals
-        fees
-        validity
-        Cardano.SNothing
-        auxiliaryDataHash
-        mint
-   where
-    noWithdrawals = Wdrl mempty
-
-instance ToJSON CardanoTxBody where
-  toJSON (Cardano.TxBody inputs outputs certificates withdrawals fees validity _update auxiliaryDataHash mint) =
-    object
-      [ "inputs" .= inputs
-      , "outputs" .= outputs
-      , "certificates" .= certificates
-      , "withdrawals" .= withdrawals
-      , "fees" .= fees
-      , "validity" .= validity
-      , "auxiliaryDataHash" .= auxiliaryDataHash
-      , "mint" .= mint
-      ]
-
-instance Crypto crypto => ToJSON (Cardano.Wdrl crypto) where
-  toJSON = toJSON . unWdrl
-
-instance Crypto crypto => FromJSON (Cardano.Wdrl crypto) where
-  parseJSON v = Cardano.Wdrl <$> parseJSON v
-
-instance ToJSON Cardano.ValidityInterval where
-  toJSON (Cardano.ValidityInterval notBefore notAfter) =
-    object
-      [ "notBefore" .= notBefore
-      , "notAfter" .= notAfter
-      ]
-
-instance FromJSON Cardano.ValidityInterval where
-  parseJSON = withObject "ValidityInterval" $ \obj ->
-    Cardano.ValidityInterval <$> obj .: "notBefore" <*> obj .: "notAfter"
-
-instance ToJSON (AuxiliaryDataHash crypto) where
-  toJSON = String . decodeUtf8 . Base16.encode . Crypto.hashToBytes . SafeHash.extractHash . unsafeAuxiliaryDataHash
-
-instance Crypto crypto => FromJSON (AuxiliaryDataHash crypto) where
-  parseJSON = withText "AuxiliaryDataHash" $ \t ->
-    case Base16.decode (encodeUtf8 t) of
-      Left e -> fail $ "decoding base16: " <> show e
-      Right bytes ->
-        case Crypto.hashFromBytes bytes of
-          Nothing -> fail "hashFromBytes yielded Nothing"
-          Just h -> pure . AuxiliaryDataHash $ SafeHash.unsafeMakeSafeHash h
+toLedgerTxId :: TxId -> Ledger.TxId Ledger.StandardCrypto
+toLedgerTxId (TxId h) =
+  Ledger.TxId (Ledger.unsafeMakeSafeHash (CC.castHash h))
+
+fromLedgerTxId :: Ledger.TxId Ledger.StandardCrypto -> TxId
+fromLedgerTxId (Ledger.TxId h) =
+  TxId (CC.castHash (Ledger.extractHash h))
 
 --
--- Input
+-- Address
 --
 
-instance Crypto crypto => ToJSON (Cardano.TxIn crypto) where
-  toJSON = toJSON . txInToText
+toLedgerAddr :: AddressInEra Era -> Ledger.Addr Ledger.StandardCrypto
```

Comment:
Could: All these functions are already in the `cardano-api` right? e.g. https://github.com/input-output-hk/cardano-node/blob/2cbe363874d0261bc62f52185cf23ed492cf4859/cardano-api/src/Cardano/Api/Address.hs#L536 for this one. Can we flag them with some comment in the code base and request access to it from upstream? Not sure if it is wise, but re-doing this feels a bit pointless.

Reply:
Partly, most of them are, some aren't exactly in the same form. But yes, ideally I would prefer not having to re-define them.